cmd/kubeadm/app/cmd/util/cmdutil.go

```diff
@@ -45,16 +45,16 @@ import (
 func SubCmdRun() func(c *cobra.Command, args []string) {
 	return func(c *cobra.Command, args []string) {
 		if len(args) > 0 {
-			kubeadmutil.CheckErr(usageErrorf(c, "invalid subcommand %q", strings.Join(args, " ")))
+			kubeadmutil.CheckErr(usageErrorf(c, "%w %q", kubeadmutil.ErrInvalidSubCommand, strings.Join(args, " ")))
```

Comment:
ditto, unclear why this is done.
if error should be part of the message it should be `%v: %q`

Reply:
We need error wrapping to make errors.Is checks instead of ugly strings.Contains here: https://github.com/kubernetes/kubernetes/commit/9c679fbd84538fbb6c18e6d809044732eb6000d9#diff-09e871502d21a7998073ec11b93c8ee78a796b56542da63499c40d9763daa60aR101

Reply:
%v means "value", %w - "wrap". I'm not just print error, I wrap it to make all other changes possible.

Reply:
there is pkg/errors.Wrap{f}()
it should be used instead of %w and fmt.Errorf()

Reply:
+1 to use pkg/errors.Wrap{f}(), so the stack trace will be included, you needn't call `errors.WithStack` to preserve the stack trace.

> // Wrap returns an error annotating err with a stack trace
> // at the point Wrap is called, and the supplied message.

Reply:
No, with errors.Wrap you will get such message:
"See 'command -h' for help and examples: invalid subcommand"

But we need:
```
invalid subcommand
See 'command -h' for help and examples
```

So only fmt.Errorf fits here.

Reply:
we should not use `%w` in kubeadm for the time being. it would mean mixing the go stdlib new errors and the legacy pkg/errors. we should not do that.

the logic here can be reworked.

we don't need to handle composite multiline errors like:
```
invalid subcommand
See 'command -h' for help and examples
```

`invalid subcommand` is the error, that we care about.
while `See 'command -h' for help and examples` is just some text that can be optionally printed if the error is `invalid subcommand`.

so remove usageError() and then:

```
// in checkErr()...

...
	case err == ErrExit:
		handleErr("", DefaultErrorExitCode)
	case strings.Contains(err.Error(), ErrInvalidSubCommandMsg):
	case errors.Is(err, ErrInvalidSubCommand):
		handleErr(err.Error(), DefaultErrorExitCode)
		// print the "See 'command -h' for help..." text here to stderr?
	default:
		switch err.(type) {

...
````

Reply:
kubeadmutil.CheckErr knows nothing about what command was called. So you can't just print `fmt.Sprintf("See '%s -h' for help and examples", c.CommandPath())`. You can say I need pass fmt'ed string into kubeadmutil.CheckErr, but this solution won't work either: in this case, we return to the previous problem, where `errors.Wrap` returns an error in a different format than the current one.

If there are no other solutions (and there really are none), sorry for wasting your time. Without fmt, you can't keep the old error format without making the code messier.

Reply:
i am not convinced there are no solutions.
if you think of something let's have a look.

but as mentioned we should not use the %w and fmt.Errorf